File: firebase/functions/src/github/sync-workflow-files.ts
import { FieldValue } from "firebase-admin/firestore";
import { HttpsError, onCall } from "firebase-functions/https";
import * as logger from "firebase-functions/logger";
import { defineSecret } from "firebase-functions/params";
import { App } from "octokit";

import { db } from "../firebase";
import {
  teamsCollectionPath,
  workflowFilesCollectionPath,
} from "../firestore-collection-paths";
import { OPENCI_DIR_QUERY, OpenciDirEntry } from "./queries";

const GITHUB_APP_ID = defineSecret("GITHUB_APP_ID");
const GITHUB_PRIVATE_KEY = defineSecret("GITHUB_PRIVATE_KEY");

/**
 * Generate a stable document ID for a workflow file.
 */
export function workflowFileDocId(
  teamId: string,
  repository: string,
  branch: string,
  fileName: string,
): string {
  return `${teamId}_${repository.replace("/", "_")}_${branch}_${fileName}`;
}

/**
 * Sync .openci/ workflow files from GitHub to Firestore.
 * Shared logic used by both webhook handler and manual sync callable.
 */
export async function syncWorkflowFilesToFirestore(params: {
  teamId: string;
  repository: string;
  branch: string;
  octokit: any;
}): Promise<{ synced: number; deleted: number }> {
  const { teamId, repository, branch, octokit } = params;
  const [owner, repo] = repository.split("/");

  const expression = `${branch}:.openci`;

  let entries: OpenciDirEntry[] = [];
  try {
    const result = await octokit.graphql(OPENCI_DIR_QUERY, {
      owner,
      repo,
      expression,
    });
    entries = (result as any).repository.object?.entries ?? [];
  } catch (e: any) {
    if (e.message?.includes("Could not resolve to an object")) {
      // .openci/ directory does not exist — delete all cached files for this repo/branch
      await deleteAllWorkflowFiles(teamId, repository, branch);
      return { synced: 0, deleted: 0 };
    }
    throw e;
  }

  const yamlEntries = entries.filter(
    (entry) =>
      entry.type === "blob" &&
      (entry.name.endsWith(".yaml") || entry.name.endsWith(".yml")) &&
      entry.object?.text,
  );

  const batch = db.batch();
  const currentFileNames = new Set<string>();
  let syncedCount = 0;

  for (const entry of yamlEntries) {
    const docId = workflowFileDocId(teamId, repository, branch, entry.name);
    const docRef = db.collection(workflowFilesCollectionPath).doc(docId);

    currentFileNames.add(entry.name);

    batch.set(
      docRef,
      {
        teamId,
        repository,
        branch,
        fileName: entry.name,
        filePath: `.openci/${entry.name}`,
        content: entry.object!.text,
        updatedAt: FieldValue.serverTimestamp(),
        syncedAt: FieldValue.serverTimestamp(),
      },
      { merge: true },
    );

    syncedCount++;
  }

  await batch.commit();

  // Delete workflow files that no longer exist in the repository
  const deletedCount = await deleteRemovedWorkflowFiles(
    teamId,
    repository,
    branch,
    currentFileNames,
  );

  logger.info(
    `Synced ${syncedCount} workflow files, deleted ${deletedCount} for ${repository}@${branch}`,
  );

  return { synced: syncedCount, deleted: deletedCount };
}

async function deleteAllWorkflowFiles(
  teamId: string,
  repository: string,
  branch: string,
): Promise<number> {
  const snapshot = await db
    .collection(workflowFilesCollectionPath)
    .where("teamId", "==", teamId)
    .where("repository", "==", repository)
    .where("branch", "==", branch)
    .get();

  if (snapshot.empty) return 0;

  const batch = db.batch();
  for (const doc of snapshot.docs) {
    batch.delete(doc.ref);
  }
  await batch.commit();

  return snapshot.size;
}

async function deleteRemovedWorkflowFiles(
  teamId: string,
  repository: string,
  branch: string,
  currentFileNames: Set<string>,
): Promise<number> {
  const snapshot = await db
    .collection(workflowFilesCollectionPath)
    .where("teamId", "==", teamId)
    .where("repository", "==", repository)
    .where("branch", "==", branch)
    .get();

  const toDelete = snapshot.docs.filter(
    (doc) => !currentFileNames.has(doc.data().fileName),
  );

  if (toDelete.length === 0) return 0;

  const batch = db.batch();
  for (const doc of toDelete) {
    batch.delete(doc.ref);
  }
  await batch.commit();

  return toDelete.length;
}

/**
 * Manual sync callable — fallback for initial setup or missed webhooks.
 */
export const syncWorkflowFiles = onCall(
  {
    region: "asia-northeast1",
    secrets: [GITHUB_APP_ID, GITHUB_PRIVATE_KEY],
  },
  async (request) => {
    if (!request.auth) {
      throw new HttpsError("unauthenticated", "Unauthenticated");
    }

    const { teamId, repository, branch } = request.data as {
      teamId: string;
      repository: string;
      branch?: string;
    };

    if (!teamId || !repository) {
      throw new HttpsError("invalid-argument", "Missing teamId or repository");
    }

    const teamRef = db.collection(teamsCollectionPath).doc(teamId);
    const teamDoc = await teamRef.get();

    if (!teamDoc.exists) {
      throw new HttpsError("not-found", "Team not found");
    }

    const teamData = teamDoc.data()!;
    const members: string[] = teamData.members || [];

    if (!members.includes(request.auth.uid)) {
      throw new HttpsError(
        "permission-denied",
        "You are not a member of this team",
      );
    }

    const installationIds = (teamData.installationIds as number[]) || [];

    if (installationIds.length === 0) {
      throw new HttpsError(
        "failed-precondition",
        "GitHub App is not installed for this team",
      );
    }

    const [owner, repo] = repository.split("/");
    const targetBranch = branch ?? "HEAD";

    try {
      const app = new App({
        appId: GITHUB_APP_ID.value(),
        privateKey: GITHUB_PRIVATE_KEY.value(),
      });

      for (const installationId of installationIds) {
        try {
          const octokit = await app.getInstallationOctokit(installationId);

          // Verify repository access
          await octokit.request("GET /repos/{owner}/{repo}", { owner, repo });

          const result = await syncWorkflowFilesToFirestore({
            teamId,
            repository,
            branch: targetBranch,
            octokit,
          });

          return result;
        } catch (e) {
          if (e instanceof HttpsError) throw e;
          continue;
        }
      }

      throw new HttpsError(
        "not-found",
        "Repository not found in any installation",
      );
    } catch (error) {
      if (error instanceof HttpsError) throw error;
      logger.error("Failed to sync workflow files", error);
      throw new HttpsError("internal", "Failed to sync workflow files");
    }
  },
);

File: firebase/dart_functions/lib/github/sync_workflow_files.dart
import 'package:google_cloud_firestore/google_cloud_firestore.dart';
import 'package:openci_shared/firestore_paths.dart';

import '../firebase.dart';
import '../util/github_urls.dart';
import '../util/logger.dart';
import 'graphql_queries.dart';
import 'installation_token.dart';
import 'workflow_parser.dart';

Future<void> syncWorkflowFiles({
  required String repository,
  required String branch,
  required int installationId,
}) async {
  final teamId = await _findTeamIdForInstallation(installationId);
  if (teamId == null) {
    logInfo('No team found for installation, skipping sync', {
      'installationId': installationId,
    });
    return;
  }

  final apiBaseUrl = await getGitHubApiBaseUrl(teamId);

  final (:token, expiresAt: _) = await getInstallationToken(
    installationId,
    apiBaseUrl: apiBaseUrl,
  );
  final dio = createGitHubDio(token, apiBaseUrl: apiBaseUrl);

  try {
    final [owner, repo] = repository.split('/');
    final expression = '$branch:.openci';

    List<OpenciDirEntry> entries;
    try {
      final response = await dio.post(
        graphqlEndpoint(apiBaseUrl),
        data: {
          'query': openciDirQuery,
          'variables': {'owner': owner, 'repo': repo, 'expression': expression},
        },
      );

      entries =
          ((response.data['data']?['repository']?['object']?['entries']
                      as List<dynamic>?) ??
                  [])
              .map((e) => OpenciDirEntry.fromJson(e as Map<String, dynamic>))
              .toList();
    } catch (e) {
      final message = e.toString();
      if (message.contains('Could not resolve to an object')) {
        await _deleteAllWorkflowFiles(teamId, repository, branch);
        return;
      }
      rethrow;
    }

    final yamlEntries = entries.where(
      (e) =>
          e.type == 'blob' &&
          (e.name.endsWith('.yaml') || e.name.endsWith('.yml')) &&
          e.text != null,
    );

    final currentFileNames = <String>{};
    var syncedCount = 0;

    for (final entry in yamlEntries) {
      final docId = workflowFileDocId(teamId, repository, branch, entry.name);
      final now = DateTime.now().toUtc().toIso8601String();

      currentFileNames.add(entry.name);

      await firestore.collection(workflowFilesCollection).doc(docId).set({
        'teamId': teamId,
        'repository': repository,
        'branch': branch,
        'fileName': entry.name,
        'filePath': '.openci/${entry.name}',
        'content': entry.text,
        'updatedAt': now,
        'syncedAt': now,
      }, options: const SetOptions.merge());

      syncedCount++;
    }

    final deletedCount = await _deleteRemovedWorkflowFiles(
      teamId,
      repository,
      branch,
      currentFileNames,
    );

    logInfo('Synced workflow files', {
      'synced': syncedCount,
      'deleted': deletedCount,
      'repo': repository,
      'branch': branch,
    });
  } catch (e, stackTrace) {
    await logError(
      'Failed to sync workflow files',
      {'repo': repository},
      e,
      stackTrace,
    );
  } finally {
    dio.close();
  }
}

Future<String?> _findTeamIdForInstallation(int installationId) async {
  try {
    final snapshot = await firestore
        .collection(teamsCollection)
        .where('installationIds', WhereFilter.arrayContains, installationId)
        .limit(1)
        .get();

    if (snapshot.docs.isEmpty) return null;
    return snapshot.docs.first.id;
  } catch (e, stackTrace) {
    await logError('Failed to find team for installation', {}, e, stackTrace);
    return null;
  }
}

Future<int> _deleteAllWorkflowFiles(
  String teamId,
  String repository,
  String branch,
) async {
  final snapshot = await firestore
      .collection(workflowFilesCollection)
      .where('teamId', WhereFilter.equal, teamId)
      .where('repository', WhereFilter.equal, repository)
      .where('branch', WhereFilter.equal, branch)
      .get();

  if (snapshot.docs.isEmpty) return 0;

  for (final doc in snapshot.docs) {
    await doc.ref.delete();
  }

  return snapshot.docs.length;
}

Future<int> _deleteRemovedWorkflowFiles(
  String teamId,
  String repository,
  String branch,
  Set<String> currentFileNames,
) async {
  final snapshot = await firestore
      .collection(workflowFilesCollection)
      .where('teamId', WhereFilter.equal, teamId)
      .where('repository', WhereFilter.equal, repository)
      .where('branch', WhereFilter.equal, branch)
      .get();

  final toDelete = snapshot.docs.where(
    (doc) => !currentFileNames.contains(doc.data()['fileName']),
  );

  var count = 0;
  for (final doc in toDelete) {
    await doc.ref.delete();
    count++;
  }

  return count;
}
